cms_coding_rules.md

```diff
@@ -0,0 +1,144 @@
+20. Pass by value arguments which are not to be modiﬁed and are built-in types or small objects; otherwise pass arguments of class types by reference or, if necessary, by pointer.
+21. Properly use rvalue references for temporary objects that will be moved.
+22. The argument to a copy constructor and to an assignment operator must be a `const` reference, while the argument for a move constructor or move assignment operator must be an rvalue reference. (\*)
+23. Do not let `const` member functions change the state of the object. Any special exceptions to this rule must still maintain thread safety. (\*)
+24. A function must never return or in any way give access to references or pointers to local variables (stack variables) outside the scope in which they are declared.
+25. Each class may have only one each of public, protected, and private sections, which must be declared in that order. (\*)
+26. Keep the ordering of methods in the header ﬁle and in the source ﬁle identical.
+27. Provide meaningful argument names in method declarations in the header ﬁle to indicate usage, unless the type fully describes the usage.
+28. Try to avoid excessively long lines of code that impair readability.
+29. Data members of a class must not be redeﬁned in derived classes since doing so hides the original data member and could create confusion between the original and redefined data members.
+
+## 5 -- Documentation Rules
+1. Always  comment  complex,  tricky,  or  non-intuitive  portions of  code.
+2. When revising code, be sure to update and revise comments.
+
+## 6 -- Packaging Rules
+#### Libraries
+1. It is discouraged to have small packages (with couple of cc and header files).
+2. Functionality used by multiple packages should go in the `.../src` and `.../interface` directories.
+3. Only header files that expose a public interface should go into the `.../interface` directory.  
+4. Include only files that are in the current directory, e.g.  
+`#include "some_header.h"`  
+ or in the `.../interface` directory, e.g  
+`#include "Subsystem/Package/interface/some_header.h"`
+#### Plugins
+5. Put plugins (e.g. EDProducers, EDAnalyzers, etc.) into a `Package/Subpackage/plugins/` directory, with its dedicated `BuildFile.xml`.
+6. Do not split plugins into header and source files. If you do split them, keep the header files in the `.../plugins` directory.
+7. All code used only by the plugins in `../plugins` should also go under `.../plugins`.
+ `plugins.cc` and `SealPlugins.cc` or any special files that just define plugins, except for template instantiations, are discouraged. 
+#### Tests
+8. Unit tests to test the functionalities of your Library and/or Plugins should go under `.../test`.
+9. Add test library/plugins in `.../test/BuildFile.xml` for the common functionality used only by your unit tests.
+10. For unit tests which simply run `cmsRun your-cfg` (to test your plugin), please use `<test name="..." command="cmsRun …"/>` in your `.../test/BuildFile.xml`.
```

Comment:
11. Unit tests should return a non 0 value from their main to indicate a test failure. Successful tests should print nothing or a very small amount to the log file. Tests should not require a human to read their output to determine if they succeed or fail.

Reply:
Should we encourage to use a unit test framework like Catch2 or CppUnit? Or does that go in too much detail?

Reply:
I addded the rule, without the detail about Catch2/CppUnit.